adam-core/src/main/scala/org/bdgenomics/adam/rdd/ShuffleRegionJoin.scala

```diff
@@ -38,488 +35,421 @@ import scala.reflect.ClassTag
-    with Serializable {
+sealed trait VictimlessSortedIntervalPartitionJoin[T, U, RT, RU]
+    extends ShuffleRegionJoin[T, U, RT, RU] {
 
-  // stores the rightIter values that might overlap the current value from the leftIter
-  private var cache: ListBuffer[(ReferenceRegion, U)] = ListBuffer.empty
-
-  protected def postProcessHits(iter: Iterator[(T, U)],
-                                currentLeft: T): Iterator[(T, RU)]
+  /**
+   * Adds elements from right to cache based on the next region
+   * encountered.
+   */
+  protected def advanceCache(setTheoryCache: SetTheoryCache[U, RT, RU],
+                             right: BufferedIterator[(ReferenceRegion, U)],
+                             until: ReferenceRegion) = {
+    while (right.hasNext &&
+      (right.head._1.compareTo(until) <= 0 || right.head._1.covers(until))) {
 
-  protected def advanceCache(until: Long): Unit = {
-    while (right.hasNext && right.head._1._1.start < until) {
       val x = right.next()
-      cache += x._1._1 -> x._2
+      setTheoryCache.cache += ((x._1, x._2))
     }
   }
 
-  protected def pruneCache(to: Long) {
-    cache = cache.dropWhile(_._1.end <= to)
-  }
-
-  protected def processHits(currentLeft: T,
-                            currentLeftRegion: ReferenceRegion): Iterator[(T, RU)] = {
-    postProcessHits(cache
-      .filter(y => {
-        y._1.overlaps(currentLeftRegion) &&
-          (y._1.start >= binRegion.start || currentLeftRegion.start >= binRegion.start)
-      })
-      .map(y => (currentLeft, y._2))
-      .toIterator, currentLeft)
-  }
-}
-
-private case class InnerSortedIntervalPartitionJoin[T: ClassTag, U: ClassTag](
-  binRegion: ReferenceRegion,
-  left: BufferedIterator[((ReferenceRegion, Int), T)],
-  right: BufferedIterator[((ReferenceRegion, Int), U)])
-    extends VictimlessSortedIntervalPartitionJoin[T, U, U] {
-
-  // no op!
-  protected def postProcessHits(iter: Iterator[(T, U)],
-                                currentLeft: T): Iterator[(T, U)] = {
-    iter
-  }
-}
-
-private case class SortedIntervalPartitionJoinAndGroupByLeft[T: ClassTag, U: ClassTag](
-  binRegion: ReferenceRegion,
-  left: BufferedIterator[((ReferenceRegion, Int), T)],
-  right: BufferedIterator[((ReferenceRegion, Int), U)])
-    extends VictimlessSortedIntervalPartitionJoin[T, U, Iterable[U]] {
-
-  protected def postProcessHits(iter: Iterator[(T, U)],
-                                currentLeft: T): Iterator[(T, Iterable[U])] = {
-    if (iter.hasNext) {
-      Iterator((currentLeft, iter.map(_._2).toIterable))
-    } else {
-      Iterator.empty
-    }
+  /**
+   * Removes elements from cache that will not be joined to any remaining tuples
+   * in left.
+   */
+  protected def pruneCache(setTheoryCache: SetTheoryCache[U, RT, RU],
+                           to: ReferenceRegion) {
+    // remove values from cache that are less that the to and that do
+    // not overlap to.
+    setTheoryCache.cache.trimStart({
+      val trimLocation =
```

Comment:
`trimStart` trims before the position given, such that when `trimLocation == 1`, index 0 is trimmed off, index 1 becomes the new start of the collection. It essentially removes the number of items passed in.